src/sailjail.c

```diff
@@ -0,0 +1,354 @@
+
+#include <sys/types.h>
+#include <unistd.h>
+
+#if !GLIB_CHECK_VERSION(2,42,0)
+#  define G_OPTION_FLAG_NONE 0
+#endif
+
+#if !GLIB_CHECK_VERSION(2,44,0)
+/*
+ * g_option_context_set_ignore_unknown_options() is definitely not
+ * a replacement for g_option_context_set_strict_posix() but at least
+ * it's a chance to test on a platform with old glib. It's not a problem
+ * with Sailfish OS which requires a much newer version of glib.
+ */
+#  define g_option_context_set_strict_posix(x, y) \
+    g_option_context_set_ignore_unknown_options(x,y)
+#endif
+
+#define DEFAULT_CONF_FILE "/etc/sailjail.conf"
+
+#define RET_CMDLINE 1
+#define RET_CONFIG 2
+#define RET_DENIED 3
+#define RET_ERR 4
+#define RET_EXEC 5
+
+typedef struct jail_args {
+    char* conf_file;
+    char* plugin_dir;
+    char* profile;
+    char* section;
+    char* sailfish_app;
+} JailArgs;
+
+static int verbose_level = 0;
+
+static
+gboolean
+jail_arg_log_output(
+    const gchar* name,
+    const gchar* value,
+    gpointer data,
+    GError** error)
+{
+    if (gutil_log_set_type(value, NULL)) {
+        return TRUE;
+    } else {
+        g_propagate_error(error, g_error_new(G_OPTION_ERROR,
+            G_OPTION_ERROR_BAD_VALUE, "Invalid log type \'%s\'", value));
+        return FALSE;
+    }
+}
+
+static
+gboolean
+jail_arg_verbose(
+    const gchar* name,
+    const gchar* value,
+    gpointer data,
+    GError** error)
+{
+    verbose_level++;
+    gutil_log_default.level = (verbose_level < 2) ?
+        GLOG_LEVEL_DEBUG : GLOG_LEVEL_VERBOSE;
+    return TRUE;
+}
+
+static
+gboolean
+jail_arg_quiet(
+    const gchar* name,
+    const gchar* value,
+    gpointer data,
+    GError** error)
+{
+    gutil_log_default.level = GLOG_LEVEL_NONE;
+    return TRUE;
+}
+
+static
+void
+jail_args_clear(
+    JailArgs* args)
+{
+    g_free(args->conf_file);
+    g_free(args->plugin_dir);
+    g_free(args->profile);
+    g_free(args->section);
+    g_free(args->sailfish_app);
+    memset(args, 0, sizeof(*args));
+}
+
+static
+GOptionContext *
+jail_opt_context_new(
+    const GOptionEntry* entries)
+{
+    GOptionContext* options = g_option_context_new("PROGRAM [ARGS...]");
+
+    g_option_context_add_main_entries(options, entries, NULL);
+    g_option_context_set_strict_posix(options, TRUE);
+    g_option_context_set_summary(options, "Runs PROGRAM in a sandbox.");
+    return options;
+}
+
+static
+int
+sailjail_main(
+    int argc,
+    char* argv[],
+    const JailConf* conf,
+    const JailArgs* args)
+{
+    int ret = RET_ERR;
+
+    /* Directory specified on the command line overrides the config */
+    SailJail* jail = jail_launch_new();
+    JailPlugins* plugins = jail_plugins_new(args->plugin_dir ?
+        args->plugin_dir : conf->plugin_dir, NULL, NULL);
+
+    if (jail_plugins_start(plugins, jail)) {
+        GError* error = NULL;
+        char* profile_path = NULL;
+        JailRules* rules;
+        JailRulesOpt opt;
+
+        memset(&opt, 0, sizeof(opt));
```

Comment:
side note: c99 & dynamic structs too can use default initializer: JailRulesOpt opt = {};  -> zero/null init that handles also those types where zero-octets does not necessarily mean zero or null vs. memset() that does not do that